File: python/quantkiller/models/monte_carlo.py
"""Monte Carlo pricing of European options under GBM.

Exact algorithm specified in contracts/rng-spec.md §5 — the loop order,
accumulation order, and antithetic pairing are part of the spec so that
all five languages produce the same price for the same seed.

    S_T = S · exp[(r - q - σ²/2)T + σ√T · Z]
    price = e^{-rT} · mean(payoffs),  std_error = e^{-rT} · √(var/n)

With antithetic=true each sample is the average of the payoffs at +Z and -Z.
"""

import math

from .. import QKError
from ..rng import Pcg32
from ._common import CALL, get_bool, get_int, get_num, get_option_type


def price(spot, strike, rate, div_yield, vol, time, option_type, paths, seed, antithetic):
    is_call = option_type == CALL
    sign = 1.0 if is_call else -1.0

    if time <= 0.0:
        raise QKError("monte_carlo_gbm requires time > 0")
    if paths < 2:
        raise QKError("monte_carlo_gbm requires paths >= 2")

    rng = Pcg32(seed)
    disc = math.exp(-rate * time)
    drift = (rate - div_yield - 0.5 * vol * vol) * time
    volt = vol * math.sqrt(time)

    total = 0.0
    total_sq = 0.0
    for _ in range(paths):
        z = rng.next_normal()
        p1 = max(sign * (spot * math.exp(drift + volt * z) - strike), 0.0)
        if antithetic:
            p2 = max(sign * (spot * math.exp(drift - volt * z) - strike), 0.0)
            s = 0.5 * (p1 + p2)
        else:
            s = p1
        total += s
        total_sq += s * s

    mean = total / paths
    var = (total_sq - paths * mean * mean) / (paths - 1)
    if var < 0.0:  # single-pass variance can go slightly negative at var ~ 0
        var = 0.0

    return {"price": disc * mean,
            "std_error": disc * math.sqrt(var / paths)}


def run(params: dict) -> dict:
    return price(
        spot=get_num(params, "spot", minimum=0.0, strict_min=True),
        strike=get_num(params, "strike", minimum=0.0, strict_min=True),
        rate=get_num(params, "rate"),
        div_yield=get_num(params, "div_yield", default=0.0),
        vol=get_num(params, "vol", minimum=0.0),
        time=get_num(params, "time", minimum=0.0),
        option_type=get_option_type(params),
        paths=get_int(params, "paths", minimum=2),
        seed=get_int(params, "seed", default=42, minimum=0),
        antithetic=get_bool(params, "antithetic", True),
    )

File: rust/quantkiller/src/models/monte_carlo.rs
//! Monte Carlo pricing of European options under GBM. Exact algorithm per
//! contracts/rng-spec.md section 5 -- loop order and accumulation order are
//! part of the spec so every language agrees for the same seed.

use std::collections::BTreeMap;

use crate::error::{QkError, QkResult};
use crate::rng::Pcg32;

#[allow(clippy::too_many_arguments)]
pub fn price(
    spot: f64, strike: f64, rate: f64, div_yield: f64, vol: f64, time: f64,
    is_call: bool, paths: u64, seed: u64, antithetic: bool,
) -> QkResult<BTreeMap<String, f64>> {
    let sign = if is_call { 1.0 } else { -1.0 };
    if time <= 0.0 {
        return Err(QkError::new("monte_carlo_gbm requires time > 0"));
    }
    if paths < 2 {
        return Err(QkError::new("monte_carlo_gbm requires paths >= 2"));
    }

    let mut rng = Pcg32::new(seed);
    let disc = (-rate * time).exp();
    let drift = (rate - div_yield - 0.5 * vol * vol) * time;
    let volt = vol * time.sqrt();

    let mut total = 0.0_f64;
    let mut total_sq = 0.0_f64;
    for _ in 0..paths {
        let z = rng.next_normal();
        let p1 = (sign * (spot * (drift + volt * z).exp() - strike)).max(0.0);
        let s = if antithetic {
            let p2 = (sign * (spot * (drift - volt * z).exp() - strike)).max(0.0);
            0.5 * (p1 + p2)
        } else {
            p1
        };
        total += s;
        total_sq += s * s;
    }

    let n = paths as f64;
    let mean = total / n;
    let mut variance = (total_sq - n * mean * mean) / (n - 1.0);
    if variance < 0.0 {
        variance = 0.0;
    }

    let mut out = BTreeMap::new();
    out.insert("price".into(), disc * mean);
    out.insert("std_error".into(), disc * (variance / n).sqrt());
    Ok(out)
}
